source/common/http/utility.cc

```diff
@@ -254,46 +254,69 @@ bool maybeAdjustForIpv6(absl::string_view absolute_url, uint64_t& offset, uint64
-    size_t equals_index = s.find('=');
-    if (equals_index == absl::string_view::npos) {
-      // The cookie is malformed if it does not have an `=`. Continue
-      // checking other cookies in this header.
-      continue;
-    }
-    absl::string_view k = s.substr(first_non_space, equals_index - first_non_space);
-    // If the key matches, parse the value from the rest of the cookie string.
-    if (k == key) {
+void forEachCookie(const HeaderMap& headers, const LowerCaseString& cookie_header,
+                   const std::function<bool (const absl::string_view&, const absl::string_view&)> cookie_consumer) {
+  const Http::HeaderMap::GetResult cookie_headers = headers.get(cookie_header);
+
+  for (size_t index = 0; index < cookie_headers.size(); index++) {
+    auto cookie_header_value = cookie_headers[index]->value().getStringView();
+
+    // Split the cookie header into individual cookies.
+    for (const auto& s : StringUtil::splitToken(cookie_header_value, ";")) {
+      // Find the key part of the cookie (i.e. the name of the cookie).
+      size_t first_non_space = s.find_first_not_of(' ');
+      size_t equals_index = s.find('=');
+      if (equals_index == absl::string_view::npos) {
+        // The cookie is malformed if it does not have an `=`. Continue
+        // checking other cookies in this header.
+        continue;
+      }
+      absl::string_view k = s.substr(first_non_space, equals_index - first_non_space);
       absl::string_view v = s.substr(equals_index + 1, s.size() - 1);
 
       // Cookie values may be wrapped in double quotes.
       // https://tools.ietf.org/html/rfc6265#section-4.1.1
       if (v.size() >= 2 && v.back() == '"' && v[0] == '"') {
         v = v.substr(1, v.size() - 2);
       }
-      return v;
+
+      bool continue_iteration = cookie_consumer(k, v);
+      if (!continue_iteration) {
+        return;
+      }
     }
   }
-  return EMPTY_STRING;
 }
 
 std::string parseCookie(const HeaderMap& headers, const std::string& key,
                         const LowerCaseString& cookie) {
-  const Http::HeaderMap::GetResult cookie_headers = headers.get(cookie);
+  std::string value;
 
-  for (size_t index = 0; index < cookie_headers.size(); index++) {
-    auto cookie_header_value = cookie_headers[index]->value().getStringView();
-    absl::string_view result = parseCookie(cookie_header_value, key);
-    if (!result.empty()) {
-      return std::string{result};
+  // Iterate over each cookie & return if its value is not empty.
+  forEachCookie(headers, cookie, [&key, &value] (const absl::string_view& k, const absl::string_view& v) -> bool {
+    if (key == k && !v.empty()) {
+      value = std::string{v};
+      return false;
     }
-  }
 
-  return EMPTY_STRING;
+    // continue iterating until a cookie that matches `key` is found.
+    return true;
+  });
+
+  return value;
+}
+
+std::map<std::string, std::string> Utility::parseCookies(const RequestHeaderMap& headers) {
```

Comment:
Better use `absl::flat_hash_map` here.

Reply:
Switched to using `absl::flat_hash_map`

For my curiosity: Wy it should be preferred over `std::map` ? I tried looking at the abseil docs and other sources but could not get a simple answer!

Reply:
Algorithmic complexity of inserts and lookups is logarithmic for `std::map`. For hashes it's basically O(1).